### PR TITLE
*: improve NULL count estimation for single column index

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve? 

An attempt to fix issue https://github.com/pingcap/tidb/issues/8563.

When column or index contains a lot of null values, our row count estimation for conditions is quite inaccurate. This PR tries to improve the estimation for column stats or single-column index stats. Note that, in https://github.com/pingcap/tidb/issues/8563, the query is on a multi-column index, and I haven't figured out a way to work in multi-column index scenarios under the current multi-column index framework.

### What is changed and how it works?

- exclude null values from histogram when building stats for single-column index;
- add additional analyze distsql task to get null count of single-column index;
- adjust logic of row count estimation for `Histogram`, `Column` and `Index`;
- ~~fix periodical panic of `show stats_histograms`, I will split this to a separate PR later;~~ split to https://github.com/pingcap/tidb/pull/9487

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - Increased code complexity

Related changes

N/A

### CODE REVIEW

File: planner/core/planbuilder.go [hunk 1]
```diff
@@ -723,20 +732,41 @@ func (b *PlanBuilder) buildAnalyzeTable(as *ast.AnalyzeTableStmt) (Plan, error)
 	return p, nil
 }
 
+// buildAuxColTasks builds auxiliary AnalyzeColumnsTask for column of single-column index,
+// because we need `NullCount` of the column histogram for single-column index histogram.
```
Comment (lamxTyler) [inline planner/core/planbuilder.go]: If you only want the null count, can we build one more cop task to scan the null count of the index? When run `analyze index`, it does not need to scan the whole table.

Comment (eurekaka) [reply planner/core/planbuilder.go]: Updated, PTAL

File: executor/analyze.go [hunk 2]
```diff
@@ -217,13 +238,45 @@ func (e *AnalyzeIndexExec) buildStats() (hist *statistics.Histogram, cms *statis
...
 	}
+	return hist, cms, nil
+}
+
+func (e *AnalyzeIndexExec) buildStats() (hist *statistics.Histogram, cms *statistics.CMSketch, err error) {
+	if err = e.open(); err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	defer func() {
+		err1 := e.result.Close()
```
Comment (lamxTyler) [inline executor/analyze.go]: Can we use `closeAll`?

File: statistics/histogram.go [hunk 3]
```diff
@@ -839,34 +838,55 @@ func (idx *Index) String() string {
 	return idx.Histogram.ToString(len(idx.Info.Columns))
 }
 
-func (idx *Index) equalRowCount(sc *stmtctx.StatementContext, b []byte, modifyCount int64) float64 {
+func (idx *Index) equalRowCount(sc *stmtctx.StatementContext, b []byte, modifyCount int64) (float64, error) {
+	if len(idx.Info.Columns) == 1 {
+		nullBytes, err := codec.EncodeKey(sc, nil, types.NewDatum(nil))
```
Comment (lamxTyler) [inline statistics/histogram.go]: What about using a constant for it?

File: statistics/table.go [hunk 4]
```diff
@@ -382,7 +382,11 @@ func (t *Table) ColumnBetweenRowCount(sc *stmtctx.StatementContext, a, b types.D
 	if !ok || c.IsInvalid(sc, t.Pseudo) {
 		return float64(t.Count) / pseudoBetweenRate
 	}
-	return c.betweenRowCount(a, b) * c.getIncreaseFactor(t.Count)
+	count := c.betweenRowCount(a, b)
+	if a.IsNull() {
```
Comment (qw4990) [inline statistics/table.go]: It's better to move this judgement into `Histogram.betweenRowCount`?

Comment (eurekaka) [reply statistics/table.go]: Hmm, `Histogram` does not know it is a column histogram or index histogram, so I add this null accounting logic in the layer of `Column` and `Index`, I think this should be clearer?

File: executor/analyze.go [hunk 5]
```diff
@@ -184,24 +195,34 @@ func (e *AnalyzeIndexExec) open() error {
 		return errors.Trace(err)
 	}
 	e.result.Fetch(ctx)
+	if len(e.idxInfo.Columns) == 1 {
+		var countNullBuilder distsql.RequestBuilder
```
Comment (lamxTyler) [inline executor/analyze.go]: Could we extract a function for them? The only difference is the range.

File: executor/analyze.go [hunk 6]
```diff
@@ -219,13 +243,36 @@ func (e *AnalyzeIndexExec) buildStats() (hist *statistics.Histogram, cms *statis
...
 				return nil, nil, errors.Trace(err)
 			}
 		}
 	}
+	return hist, cms, nil
+}
+
+func (e *AnalyzeIndexExec) buildStats() (hist *statistics.Histogram, cms *statistics.CMSketch, err error) {
+	if err = e.open(); err != nil {
+		return nil, nil, errors.Trace(err)
```
Comment (winoros) [inline executor/analyze.go]: remove trace.

File: executor/analyze.go [hunk 7]
```diff
@@ -181,29 +183,51 @@ func (e *AnalyzeIndexExec) open() error {
...
-			hist = nil
-			cms = nil
-			err = errors.Trace(err1)
+	err := e.fetchAnalyzeResult(ranges, false)
+	if err != nil {
+		return err
+	}
+	if len(e.idxInfo.Columns) == 1 {
+		ranges = ranger.NullRange()
+		err = e.fetchAnalyzeResult(ranges, true)
```
[the remark this answers is not in the document]
Comment (eurekaka) [reply executor/analyze.go]: Yes, we would call distsql twice for single column index now.

**Comment by zz-jason** [comment]
@eurekaka please cherry pick this PR to release-2.1

**Pull request merged**